### ISSUE TITLE
How to handle custom events in existing Widget?

### ISSUE THREAD

**Issue by sum-elier** [hecrj/iced#356]
The `TextInput` widget has already a bunch of logic inside of the `on_event` function from the `Widget` trait. I would like to be able to extend this functionality by having a custom event handler such that it is executed first, before any of the cases inside the `match` inside `on_event`, such that the custom event handler returns `true` or `false` indicating whether the event was consumed or not.

My immediate use case is to be able to jump between multiple `TextInput` by using the Tab and Shift+Tab keys, in an arbitrary order.

Is there any way to achieve this?

**Comment by Imberflur** [comment]
You could create a custom container widget that filters events before passing them to the `TextInput`.

**Comment by hecrj** [comment]
As @Imberflur pointed out, you can build a custom container widget to manage text input focus.

In the long run, we could include a `Form` widget of sorts, supporting tab-focusing out of the box. I am the author of [`composable-form`](https://github.com/hecrj/composable-form) for Elm and I'd love to see a similar project for `iced`!

Feel free to share any ideas, use cases, or suggestions.

**Comment by sum-elier** [comment]
> 
> 
> You could create a custom container widget that filters events before passing them to the `TextInput`.

I am not sure if that would help because I am not sure if it is possible to give focus to another widget. I see there's a `is_focused` method in `text_input::State` but that's it. Can't find any other indication on how to give focus to another widget. Which module should I be searching for that functionality?

**Comment by Imberflur** [comment]
@sum-elier 
You can use the [`text_input::State::focused()`](https://docs.rs/iced_native/0.2.2/iced_native/widget/text_input/struct.State.html#method.focused) method to create a new `State` that is focused to replace the old one.

**Comment by sum-elier** [comment]
@Imberflur 
I am thinking I've must designed my solution wrong or something, but it seems to me there's no way to achieve this since one usually has a struct with all the `State` fields, as in `struct Controls` in Iced's todos/src/main.rs demo, and then inside the `fn view(&mut self)` method one passes the state as `&mut self.text_input_1_state` to the `new` method of the widget, for example `TextInput`.
Once the state has been passed in there's no way to change it in response to a specific event since it has already been exclusively borrowed by `new`. The way I am filtering events before passing the to `TextInput` is by using a closure in a wrapper around `TextInput`. Since the closure will also modify the state to update it to not focused after detecting a tab, that would cause a compilation error.
Is that correct?

**Comment by Imberflur** [comment]
@sum-elier 
That's correct. I believe the idiomatic way to do this in `iced` is generating a `Message` in your custom widget and handling mutation in your `update` method. Generally, the closures passed to widgets are not meant to have side-effects and are instead meant to transform whatever input the widget feeds it into a `Message`.

**Comment by sum-elier** [comment]
@Imberflur 
The thing is that the custom widget only wraps around `TextInput` thus there's no way to change `text_input::State` once the exclusive reference to it has been passed into `TextInput::new()` so I am not sure how to accomplish what you mention. The only solution I see is copying all the code from `TextInput` and modifying the implementation of `on_event`.

**Comment by Imberflur** [comment]
I am suggesting you do the mutation of `text_input::State` once the exclusive reference is released. So instead of inside your `view` method or inside your wrapping widget's methods it would be in your `update` method.
